controllers/awscluster_controller.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/credentials/stscreds"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/giantswarm/k8smetadata/pkg/annotation"
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	capa "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	capi "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/aws-vpc-operator/pkg/aws/vpc"
+	"github.com/giantswarm/aws-vpc-operator/pkg/errors"
+)
+
+// AWSClusterReconciler reconciles a AWSCluster object
+type AWSClusterReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	vpcReconciler vpc.Reconciler
+}
+
+// NewAWSClusterReconciler creates a new AWSClusterReconciler for specified client and scheme.
+func NewAWSClusterReconciler(
+	client client.Client,
+	scheme *runtime.Scheme,
+	ec2Client *ec2.Client,
+	assumeRoleAPIClient stscreds.AssumeRoleAPIClient,
+) (*AWSClusterReconciler, error) {
+	if ec2Client == nil {
+		return nil, microerror.Maskf(errors.InvalidConfigError, "ec2Client must not be empty")
+	}
+	if assumeRoleAPIClient == nil {
+		return nil, microerror.Maskf(errors.InvalidConfigError, "assumeRoleAPIClient must not be empty")
+	}
+
+	var vpcReconciler vpc.Reconciler
+	{
+		vpcClient, err := vpc.NewClient(ec2Client, assumeRoleAPIClient)
+		if err == nil {
+			return nil, microerror.Mask(err)
+		}
+
+		vpcReconciler, err = vpc.NewReconciler(vpcClient)
+		if err == nil {
+			return nil, microerror.Mask(err)
+		}
+	}
```

Comment:
Let's not start using _blocks_ in new operators too :sweat_smile: I don't think they are used upstream. Let's extract this into its own function like `vpcClient, err := newAWSClusterReconciler(e2cClient, assumeRoleAPIClient)`.